src/Common/src/System/Runtime/InteropServices/StringBuffer.cs

```diff
@@ -0,0 +1,385 @@
+        /// Instantiate the buffer with a copy of the specified string.
+        /// </summary>
+        public unsafe StringBuffer(string initialContents)
+            : base(0)
+        {
+            // We don't pass the count of bytes to the base constructor, appending will
+            // initialize to the correct size for the specified initial contents.
+            if (initialContents != null)
+            {
+                this.Append(initialContents);
+            }
+        }
+
+        /// <summary>
+        /// Get/set the character at the given index.
+        /// </summary>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown if attempting to index outside of the buffer length.</exception>
+        public new unsafe char this[ulong index]
+        {
+            get
+            {
+                if (index >= _length) throw new ArgumentOutOfRangeException("index");
+                return CharPointer[index];
+            }
+            set
+            {
+                if (index >= _length) throw new ArgumentOutOfRangeException("index");
+                CharPointer[index] = value;
+            }
+        }
+
+        /// <summary>
+        /// Character capacity of the buffer. Includes the count for the trailing null character.
+        /// </summary>
+        public ulong CharCapacity
+        {
+            get
+            {
+                ulong byteCapacity = ByteCapacity;
+                return byteCapacity == 0 ? 0 : byteCapacity / sizeof(char);
+            }
+        }
+
+        /// <summary>
+        /// Ensure capacity in characters is at least the given minimum.
+        /// </summary>
+        /// <exception cref="OutOfMemoryException">Thrown if unable to allocate memory when setting.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown if attempting to set <paramref name="Capacity"/> to a value that is larger than the maximum addressable memory.</exception>
+        public override void EnsureCapacity(ulong minCapacity)
+        {
+            if (minCapacity > (ulong.MaxValue / sizeof(char))) throw new ArgumentOutOfRangeException("minCapacity");
+            base.EnsureCapacity(minCapacity * sizeof(char));
+        }
+
+        /// <summary>
+        /// The logical length of the buffer in characters. (Does not include the final null.) Will automatically attempt to increase capacity.
+        /// This is where the usable data ends.
+        /// </summary>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown if attempting to set <paramref name="nameof(Length)"/> to a value that is larger than the maximum addressable memory.</exception>
+        /// <exception cref="OutOfMemoryException">Thrown if unable to allocate memory when setting.</exception>
+        public unsafe ulong Length
+        {
+            get { return _length; }
+            set
+            {
+                // Null terminate
+                EnsureCapacity(value + 1);
```

Comment:
I don't see it checking this.  If you set Length to -5, Ensure capacity is going to compare -4 to see whether it's smaller than its internal capacity, which it will be, so it'll be a nop, and then we'll index into CharPointer at -4.  Am I missing where it's being checked?